tests/recordCallsToSetTimeout.ts

```diff
@@ -0,0 +1,193 @@
+  }: {
+    originalSetTimeout: OriginalSetTimeout;
+    numAutomaticCalls?: number;
+    interceptCallback?: InterceptingCallback;
+  }) {
+    this.originalSetTimeout = originalSetTimeout;
+    this.interceptCallback = interceptCallback;
+
+    this.calls = [];
+    this.events = new EventEmitter();
+    this.numAutomaticCallsRemaining = numAutomaticCalls;
+  }
+
+  /**
+   * Looks for the first `setTimeout` call in the call stack that matches the
+   * given duration and calls it, removing it from the call stack, or waits
+   * until such a call appears.
+   *
+   * @param duration - The expected duration of a `setTimeout` call.
+   * @returns A promise that resolves when a `setTimeout` call matching the
+   * given duration is called.
+   */
+  nextMatchingDuration(duration: number): Promise<void> {
+    return new Promise<void>((resolve) => {
+      const index = this.calls.findIndex((call) => call.duration === duration);
+
+      if (index === -1) {
+        const listener = (call: SetTimeoutCall, callIndex: number) => {
+          if (call.duration === duration) {
+            call.callback();
+            this.calls.splice(callIndex, 1);
+            this.events.off('setTimeoutAdded', listener);
+            resolve();
+          }
+        };
+        this.events.on('setTimeoutAdded', listener);
+      } else {
+        this.calls[index].callback();
+        this.calls.splice(index, 1);
+        resolve();
+      }
+    });
+  }
+
+  /**
+   * Registers a callback that will be called when `setTimeout` is called and
+   * the expected number of `setTimeout` calls (as specified via
+   * `numAutomaticCalls`) is exceeded.
+   *
+   * @param callback - The callback to register.
+   */
+  onNumAutomaticCallsExhausted(callback: () => void): void {
+    this.events.on('numCallsToPassThroughExhausted', callback);
+  }
+
+  /**
+   * The function with which to replace the global `setTimeout` function. This
+   * mock implementation will record the call to `setTimeout`, along with its
+   * callback and duration, in a call stack, which can be accessed later.
+   *
+   * @param callback - The callback associated with a particular `setTimeout`
+   * call.
+   * @param duration - The duration associated with a particular `setTimeout`
+   * call.
+   * @returns An instance of NodeJS.Timeout which is only supplied to fulfill
+   * the existing type of `setTimeout` and serves no purpose.
+   */
+  _mockSetTimeoutImplementation = (
+    callback: SetTimeoutCallback,
+    duration: number | undefined = 0,
+  ): NodeJS.Timeout => {
+    // Build a "fake" Timeout object
+    const timeout = this.originalSetTimeout(EMPTY_FUNCTION, 0);
+    const interceptedCallback = this.interceptCallback(
+      callback,
+      this._stopPassingThroughCalls,
+    );
+    const call = {
+      callback: interceptedCallback,
+      duration,
+      timeout,
+    };
+    this.calls.push(call);
+
+    if (this.numAutomaticCallsRemaining > 0) {
```

Comment:
This is one of the reasons why I had to write my own mock logic. This setting tells the SetTimeoutRecorder, "if you see a `setTimeout`, go ahead and call its callback immediately — but only do this a certain amount of times". This helps in writing tests because:

1. This strategy works with `setTimeout`s that have yet to take place, unlike Jest's fake timers, where you can only call `setTimeout`s that have happened in the past.
2. You can designate a maximum number of `setTimeout` calls in your test. This means if you have a situation where `setTimeout` is called in a loop, as is the case with PollingBlockTracker, you can't end up in an infinite loop, which happened to me more times than I liked while using Jest's fake timers.

Reply:
I'm having trouble understanding the scenario where an infinite loop happens when using fake timers. I also don't really understand what you mean about only being able to call `setTimeout`s that "happen in the past".

Reply:
PollingBlockTracker contains a loop that will periodically fetch the latest block from the network and emit the `latest` event when the block changes. The periodicity is implemented using `setTimeout`:

``` typescript
async _start() {
  while (this._isRunning) {
    await fetchLatestBlock();   // condensed version
    await new Promise((resolve) => {
      setTimeout(resolve, 20000);
    })
  }
}
```

This pattern, combined with the fact that the promise that `_start()` returns is never `await`ed, means that stubbing `setTimeout` is tricky. Jest's fake timers keeps track of `setTimeout` calls that have occurred in the past, but not in the future. To illustrate this, let's say you start writing a test that looks like this:

``` typescript
jest.useFakeTimers();
const blockTracker = new PollingBlockTracker(...);
const receivedBlockNumbers: string[] = [];
await new Promise((resolve) => {
  // start the block tracker by listening to the event
  blockTracker.on('latest', (blockNumber: string) => {
    receivedBlockNumbers.push(blockNumber);
    if (receivedBlockNumbers.length === 3) {
      resolve();
    }
  });
})
```

If something is stubbing `setTimeout`, then the promise within the loop in `_start()` will never resolve. How do you resolve it? By calling the callback that `setTimeout` takes. But where do you do that? Maybe you could do it after `latest` is listened to:

``` typescript
jest.useFakeTimers();
const blockTracker = new PollingBlockTracker(...);
const receivedBlockNumbers: string[] = [];
const promiseForLatestBlocks = new Promise((resolve) => {
  // start the block tracker by listening to the event
  blockTracker.on('latest', (blockNumber: string) => {
    receivedBlockNumbers.push(blockNumber);
    if (receivedBlockNumbers.length === 3) {
      resolve();
    }
  });
});
jest.runOnlyPendingTimers();
await promiseForLatestBlocks;
```

But this won't work, because `setTimeout` isn't called right after `latest` is listened to; it's called a little bit after. You can't even wait until `latest` is emitted, because that's too early as well. 

How I ended up solving this problem later on, after I'd developed my stub for `setTimeout`, was to add an event, `_waitingForNextIteration`, that would be emitted right after `setTimeout` was called. So, now that this event is in place, I think I could solve this problem using Jest's fake timers by saying:

``` typescript
jest.useFakeTimers();
const blockTracker = new PollingBlockTracker(...);
const receivedBlockNumbers: string[] = [];
blockTracker.on('_waitingForNextIteration', () => {
  if (receivedBlockNumbers.length < 3) {
    jest.runOnlyPendingTimers();
  }
});
await new Promise((resolve) => {
  // start the block tracker by listening to the event
  blockTracker.on('latest', (blockNumber: string) => {
    receivedBlockNumbers.push(blockNumber);
    if (receivedBlockNumbers.length === 3) {
      resolve();
    }
  });
});
```

Note that `if (receivedBlockNumbers.length < 3) {` check is key in the `_waitingForNextIteration` listener, because without it, as soon as `setTimeout` is called, its callback will be run. Because this controls the loop, leaving this out also would cause the loop to run indefinitely.

Reply:
Per our conversation in standup today, I'm going to leave these tests using my stubs for `setTimeout` / `clearTimeout` so I can move onto other things, but I did make an issue for converting them to Jest's fake timers here: https://github.com/MetaMask/eth-block-tracker/pull/103